## Simpson's Rule

In [1]:
import numpy as np
import matplotlib.pyplot as plt

### Definition
Simpson's rule uses a quadratic polynomial on each subinterval of a partition to approximate the function $f(x)$ and to compute the definite integral. This is an improvement over the trapezoid rule which approximate $f(x)$ by a straight line on each subinterval of a partition.\\
The formula for Simpson's rule is $$S_{N}f = \frac{\Delta}{3}\sum_{i=1}^{N/2}\left(f(x_{2i-2})+4f(x_{2i-1}+f(x_{2i})\right)$$
where $N$ is an even number of subintervals of $\left[a, b \right]$, $\Delta x =(b-a)/N$ and $x_{i}=a+i\Delta x$.

### Error Formula
We have seen that the error in a Riemann sum is inversely proportional to the size of the partition $N$ and the trapezoid rule is inversely proportional to $N^{2}$. The error formula in the theorem below shows that Simpson's rule is even better as the error is inversely proportional $N^{4}$.

**Theorem** Let $S_{N}(f)$ denote Simpson's rule $$S_{N}f = \frac{\Delta}{3}\sum_{i=1}^{N/2}\left(f(x_{2i-2})+4f(x_{2i-1}+f(x_{2i})\right)$$ where $N$ is an even number of subintervals of
$\left[a, b\right]$, $\Delta x = (b-a)/N$ and $x_{i}=a+i\Delta x$. The error bound is
$$E_{N}^{S}(f)=\left|\int_{a}^{b}f(x)dx - S_{N}(f)\right| \leq \frac{(b-a)^{5}}{180N^{4}}k_{4}$$
where $\left|f^{(4)}\right| \leq K_{4}$ for all $x \in \left[a, b\right]$.

### Implementation
Let's write a function called `simps` which takes input parameters $f, a, b$ and $N$ and returns the approximatoin $S_{N}(f)$. Furthermore, let's assign a default value $N=50$.

In [4]:
x = np.linspace(0, 1, 11)
print(x)
print(x[0:-1:2])
print(x[1::2])
print(x[2::2])

[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1. ]
[0.  0.2 0.4 0.6 0.8]
[0.1 0.3 0.5 0.7 0.9]
[0.2 0.4 0.6 0.8 1. ]


In [5]:
def simps(f, a, b, N=50):
    if N % 2 ==1:
        raise ValueError("N must be an even integer.")
    dx = (b-a)/N
    x = np.linspace(a, b, N+1)
    y = f(x)
    S = dx/3 * np.sum(y[0:-1:2] + 4*y[1::2] + y[2::2])
    return S

Let's test our function on integrals for which we know te exact value. For example, we know $$\int_{0}^{1}3x^{2}dx=1$$

In [6]:
simps(lambda x : 3*x**2, 0, 1, 10)

1.0

Test our function again with the integral $$\int_{0}^{\pi/2}\sin{x}dx=1$$

In [7]:
simps(np.sin, 0, np.pi/2, 100)

1.000000000338236

#### Examples
##### Approximate ln(2)
Find a value $N$ which guarantees that Simpson's rule approximate $S_{N}(f)$ of the integral $$\int_{1}^{2}\frac{1}{x}dx$$
satisfies $E_{N}^{S}(f) \leq 0.0001$.
Compute $$f^{(4)} = \frac{24}{x^{5}}$$
therefore $\left| f^{(4)} \right|\leq 24$ for all $x \in \left[1, 2\right]$ and so
$$\frac{1}{180N^{4}}24\leq 0.0001 \Rightarrow \frac{20000}{15N^{4}} \leq 1 \Rightarrow \left(\frac{20000}{15}\right)^{1/4} \leq N$$
Compute

In [8]:
(20000/15)**0.25

6.042750794713537

Compute Simpson's rule with $N=8$(the smallest even integer greater than 6.04)

In [11]:
approximation = simps(lambda x: 1/x, 1, 2, 8)
print(approximation)

0.6931545306545306


Verify that $E_{N}^{S}(f) \leq 0.0001$

In [12]:
np.abs(np.log(2)-approximation)

7.350094585301115e-06